torch/csrc/jit/graph_executor.cpp

```diff
@@ -123,48 +138,37 @@ struct ExecutionPlan {
   }
 
 private:
-  // note: should be inplace to avoid allocations, but we have to switch from
-  // a list of tensor to a list of ivalues
-  std::vector<IValue> unwrapVariables(variable_tensor_list && list) const {
-    return fmap(list, [](const Variable& v) -> IValue {
-      return v.defined() ? autograd::as_variable_ref(v).detach() : at::Tensor();
-    });
-  }
-  // note: should be inplace to avoid allocations, but we have to switch from
-  // a list of tensor to a list of ivalues
-  variable_tensor_list wrapTensors(tensor_list && list) const {
-    for(auto & v : list) {
-      v = autograd::make_variable(v, /*requires_grad=*/false);
+  void detachVariables(Stack & stack) const {
+    for (IValue & v : stack) {
+      if (!v.isTensor()) continue;
+      auto t = std::move(v).toTensor();
+      v = IValue{t.defined() ? autograd::as_variable_ref(t).detach() : std::move(t)};
     }
-    return variable_tensor_list(std::move(list));
   }
   // Capture (save) inputs that would be required to subsequently run backwards
-  void captureInputs(ExecutionPlanAutogradFunction & grad_fn, variable_tensor_list & inputs) const {
-    for(auto offset : grad.df_input_captured_inputs) {
-      grad_fn.captures.emplace_back(autograd::as_variable_ref(inputs[offset]), false);
+  void captureInputs(ExecutionPlanAutogradFunction & grad_fn, Stack & inputs) const {
+    for (size_t offset : grad.df_input_captured_inputs) {
+      grad_fn.capture(inputs[offset]);
     }
   }
-  void captureOutputs(ExecutionPlanAutogradFunction & grad_fn, variable_tensor_list & outputs) const {
-    for(auto offset : grad.df_input_captured_outputs) {
-      grad_fn.captures.emplace_back(autograd::as_variable_ref(outputs[offset]), true);
+  void captureOutputs(ExecutionPlanAutogradFunction & grad_fn, Stack & outputs) const {
+    for (size_t offset : grad.df_input_captured_outputs) {
+      grad_fn.capture(outputs[offset]);
     }
   }
 
-  variable_tensor_list runWithGrad(variable_tensor_list&& inputs) const {
+  void runWithGrad(Stack & stack) const {
     auto grad_fn = std::make_shared<ExecutionPlanAutogradFunction>(grad_executor,
       grad.df_input_captured_inputs.size() + grad.df_input_captured_outputs.size());
-    // hook up the outputs of df to the gradient functions of the inputs that require
-    // gradients
+    // hook up the outputs of df to the gradient functions of the inputs that require gradients
     for(auto idx : grad.df_output_vjps) {
-      auto & v = autograd::as_variable_ref(inputs[idx]);
+      auto v = Variable(stack[idx].toTensor());
       grad_fn->add_next_edge(v.gradient_edge());
     }
-    captureInputs(*grad_fn, inputs);
+    captureInputs(*grad_fn, stack);
 
-    auto stack = unwrapVariables(std::move(inputs));
+    detachVariables(stack);
```

Comment:
In all other uses of stacks, the assumption is that the stack can hold more inputs that what the operator will consume. Here it is adding a strong assumption that the stack only contains the inputs for this executor. It would be safer if we only examined the top N elements as determined by the graph. 

We do not want to guarantee that the stack has to be empty except for the current arguments, because it precludes a lot of stack movement optimizations.

Reply:
Cool, I'll limit those functions to only operate on a subset of the stack.